saivpp/src/SwitchStateBaseAcl.cpp

```diff
@@ -461,202 +604,458 @@ sai_status_t acl_priority_attr_get (
+    } else {
+        return status;
+    }
 
-		    auto ace_it = m_ace_cntr_info_map.find(ace_cntr_oid);
+    std::list<std::string> hwif_names;
+    SWSS_LOG_INFO("Adding tunterm acl rules, tunterm_acl index %u", tunterm_acl_swindex);
 
-		    if (ace_it != m_ace_cntr_info_map.end()) {
-			m_ace_cntr_info_map.erase(ace_it);
-		    }
-	
-		    // For stats we need to find vpp rule index from acl_entry_counter (ace_counter)
-		    m_ace_cntr_info_map[ace_cntr_oid] = { tbl_oid, ace.ace_oid, acl_swindex, index };
-		}
-	    }
-	    index++;
-	}
+    status = vpp_tunterm_acl_add_replace(&tunterm_acl_swindex, acl->count, acl);
+
+    if (status == SAI_STATUS_SUCCESS) {
+        m_tunterm_acl_swindex_map[tbl_oid] = tunterm_acl_swindex;
+    } else {
+        return SAI_STATUS_FAILURE;
+    }
+
+    if (do_port_bind) {
+        auto it_hw_ports = m_acl_tbl_hw_ports_map.find(tbl_oid);
+        if (it_hw_ports == m_acl_tbl_hw_ports_map.end()) {
+            auto sid = sai_serialize_object_id(tbl_oid);
+            SWSS_LOG_INFO("Tunterm acl - no ports bound for table id %s", sid.c_str());
+        } else {
+            hwif_names = it_hw_ports->second;
+            for (auto hwif_name: hwif_names) {
+                SWSS_LOG_INFO("Tunterm acl - binding to hwif %s", hwif_name.c_str());
+                tunterm_acl_bindunbind(tbl_oid, true, hwif_name);
+            }
+        }
     }
-    SWSS_LOG_NOTICE("ACL table %s %s status %d", tbl_sid.c_str(),
-		    acl_replace ? "replace" : "add", status);
-    free(aces);
-    free(acl);
-    ordered_aces.clear();
 
+    SWSS_LOG_INFO("Tunterm ACL table %s %s status %d", tbl_sid.c_str(),
+            do_port_bind ? "add" : "replace", status);
+    
+    return status;
+}
+
+sai_status_t SwitchStateBase::tbl_hw_ports_map_delete(sai_object_id_t tbl_oid)
+{
+    auto hw_ports_it = m_acl_tbl_hw_ports_map.find(tbl_oid);
+    if (hw_ports_it == m_acl_tbl_hw_ports_map.end()) {
+        return SAI_STATUS_SUCCESS;
+    }
+    m_acl_tbl_hw_ports_map.erase(hw_ports_it);
+    return SAI_STATUS_SUCCESS;
+}
+
+sai_status_t SwitchStateBase::tunterm_acl_delete(sai_object_id_t tbl_oid, bool table_delete)
+{
+    sai_status_t status = SAI_STATUS_SUCCESS;
+    auto         tunterm_idx_it = m_tunterm_acl_swindex_map.find(tbl_oid);
+
+    if (tunterm_idx_it == m_tunterm_acl_swindex_map.end()) {
+        SWSS_LOG_WARN("No tunterm ACL configured for table %s",
+                        sai_serialize_object_id(tbl_oid).c_str());
+        return status;
+    } 
+
+    if(!table_delete) {
+        for(auto hwif_name:  m_acl_tbl_hw_ports_map[tbl_oid]) {
+            tunterm_acl_bindunbind(tbl_oid, false, hwif_name);
+        }
+    }
+
+    uint32_t tunterm_acl_swindex = tunterm_idx_it->second;
+    status = vpp_tunterm_acl_del(tunterm_acl_swindex);
+    if (status == SAI_STATUS_SUCCESS) {
+        m_tunterm_acl_swindex_map.erase(tunterm_idx_it);      
+    }
+    
+    if(table_delete && (status == SAI_STATUS_SUCCESS)) {
```

Comment:
You've already checked STATUS_SUCCESS above, you can move this check on table_delete inside the same block or return on non success.

Reply:
Updated